pkg/recipes/terraform/config/config.go

```diff
@@ -190,23 +237,26 @@ func getProviderConfigs(ctx context.Context, requiredProviders []string, ucpConf
 			return nil, err
 		}
 		if len(config) > 0 {
-			providerConfigs[provider] = config
+			providerConfigs[providerName] = config
 		}
 	}
 
 	return providerConfigs, nil
 }
 
-// AddTerraformBackend adds backend configurations to store Terraform state file for the deployment.
+// AddTerraformInfrastructure adds backend configurations to store Terraform state file for the deployment.
+// It also sets the required providers for the Terraform configuration.
 // Save() must be called to save the generated backend config.
 // Currently, the supported backend for Terraform Recipes is Kubernetes secret. https://developer.hashicorp.com/terraform/language/settings/backends/kubernetes
-func (cfg *TerraformConfig) AddTerraformBackend(resourceRecipe *recipes.ResourceMetadata, backend backends.Backend) (map[string]any, error) {
+func (cfg *TerraformConfig) AddTerraformInfrastructure(resourceRecipe *recipes.ResourceMetadata, backend backends.Backend, requiredProviders map[string]*RequiredProviderInfo) (map[string]any, error) {
 	backendConfig, err := backend.BuildBackend(resourceRecipe)
 	if err != nil {
 		return nil, err
 	}
+
 	cfg.Terraform = &TerraformDefinition{
-		Backend: backendConfig,
+		Backend:           backendConfig,
+		RequiredProviders: requiredProviders,
```

Comment:
What's the motivation of adding this with backend instead of where providers are added?

Reply:
Terraform expects the required_providers configuration to be included with backend. This is needed in order to ensure processing of providers which it does not recognize having a default configuration.
It is also beneficial for providers which Terraform may use default configuration for, so we use the user-intended required_provider configuration instead of default versions.

Reply:
I understand required_providers need to be added to the config. My question is the code path where you are adding this. We shouldn't be overloading the backend config generation with providers, all the providers stuff should be processed in a single place.

Reply:
Removed updates to this function as discussed. All work of adding provider configuration is now in AddProviders() function